src/bitcoin.js

```diff
@@ -345,25 +349,82 @@ function writeUInt64LE(n, buf, off) {
     try {
       const bech32Dec = bech32.decode(address);
-      if (bech32Dec.prefix === 'bc')
-        versionByte = 0xD0;
-      else if (bech32Dec.prefix === 'tb')
-        versionByte = 0xF0;
+      if (bech32Dec.prefix === SEGWIT_NATIVE_V0_PREFIX)
+        versionByte = FMT_SEGWIT_NATIVE_V0;
+      else if (bech32Dec.prefix === SEGWIT_NATIVE_V0_TESTNET_PREFIX)
+        versionByte = FMT_SEGWIT_NATIVE_V0_TESTNET;
       else
         throw new Error('Unsupported prefix: must be bc or tb.');
       if (bech32Dec.words[0] !== 0)
         throw new Error(`Unsupported segwit version: must be 0, got ${bech32Dec.words[0]}`);
+      
       pkh = Buffer.from(bech32.fromWords(bech32Dec.words.slice(1)));
     } catch (err) {
       throw new Error(`Unable to decode address: ${address}: ${err.message}`)
     }
   }
   return {versionByte, pkh};
+}
+
+// Determine the address format (a.k.a. "version") depending on the
+// purpose of the dervation path.
+function getAddressFormat(path) {
+  if (path.length < 2)
+    throw new Error('Path must be >1 index');
+  const purpose = path[0];
+  const coin = path[1];
+  if (purpose === BTC_PURPOSE_P2WPKH && coin === BTC_COIN) {
+    return FMT_SEGWIT_NATIVE_V0;
+  } else if (purpose === BTC_PURPOSE_P2WPKH && coin === BTC_TESTNET_COIN) {
+    return FMT_SEGWIT_NATIVE_V0_TESTNET;
+  } else if (purpose === BTC_PURPOSE_P2SH_P2WPKH && coin === BTC_COIN) {
+    return FMT_SEGWIT_WRAPPED;
+  } else if (purpose === BTC_PURPOSE_P2SH_P2WPKH && coin === BTC_TESTNET_COIN) {
+    return FMT_SEGWIT_WRAPPED_TESTNET;
+  } else if (purpose === BTC_LEGACY_PURPOSE && coin === BTC_COIN) {
+    return FMT_LEGACY;
+  } else if (purpose === BTC_LEGACY_PURPOSE && coin === BTC_TESTNET_COIN) {
+    return FMT_LEGACY_TESTNET;
+  } else {
+    throw new Error('Invalid Bitcoin path provided. Cannot determine address format.')
+  }
+}
+exports.getAddressFormat = getAddressFormat;
+
+// Determine the script type for an input based on its owner's derivation
+// path's `purpose` index. 
+// We do not support p2sh and only issue single-key addresses from the Lattice
+// so we can determine this based on path alone.
+function getScriptType(input) {
+  switch (input.signerPath[0]) {
+    case BTC_LEGACY_PURPOSE:
+      return BTC_SCRIPT_TYPE_P2PKH;
+    case BTC_PURPOSE_P2SH_P2WPKH:
+      return BTC_SCRIPT_TYPE_P2SH_P2WPKH;
+    case BTC_PURPOSE_P2WPKH:
+      return BTC_SCRIPT_TYPE_P2WPKH_V0;
+    default:
+      throw new Error(`Unsupported path purpose (${input.signerPath[0]}): cannot determine BTC script type.`);
+  }
+}
+
+// Determine if a a transaction should have a witness portion.
+// This will return true if any input is p2sh(p2wpkh) or p2wpkh.
+// We determine the script type based on the derivation path.
+function needsWitness(inputs) {
+  let w = false;
+  inputs.forEach((input) => {
+    if ((input.signerPath[0] === BTC_PURPOSE_P2WPKH) || 
+        (input.signerPath[0] === BTC_PURPOSE_P2SH_P2WPKH)) {
+      w = true;
```

Comment:
Sorry for my JS n00b, but if it's possible to `break` a `forEach`, it could be done here...

Reply:
Breaking doesn't work very well in JS. Neither does returning `true` here. There are weird rules that i've never been able to figure out